stripped/colabs/Analiza_szeregow_czasowych_dot_roslinnosci_stripped.py

```diff
@@ -0,0 +1,180 @@
+
+
+"""
+Zanim zacznimy pisać kwerendy należy jeszcze dodać nasz DataFrame (df) do "przestrzeni nazw tabel" Sparka:
+"""
+
+nasa.createOrReplaceTempView("nasa")
+
+
+"""
+Rozdzielenie kolumny "Date" na kolumny "Year" oraz "Month"
+"""
+
+nasa = spark.sql("""
+          SELECT
+          CAST(SUBSTRING(CAST(Date AS STRING), 1, 4) AS INT) AS Year,
+          CAST(SUBSTRING(CAST(Date AS STRING), 5, 2) AS INT) AS Month,
+          n.*
+          FROM nasa n
+          """)
+
+
+nasa = nasa.drop("Date")
+nasa.createOrReplaceTempView("nasa")
+
+
+"""
+# Analiza szergów czasowych  dot. roślinności
+"""
+
+# Import biblotek
+import matplotlib.pyplot as plt
+import random
+import matplotlib.gridspec as gridspec
+
+
+# Ograniczenie zbioru nasa do wartości, które potrzebujemy, ustawienie pomiaru zawsze na 1 dzień miesiąca
+szeregNasa = spark.sql("""
+                      SELECT
+                      lon, lat,
+                      to_date(CONCAT(Year, '-', Month, '-1')) as Date, GVEG, LAI
+                      FROM nasa
+                      order by lon, lat, Date
+                      """)
+
+
+%%time
+szeregNasa.show(5)
+
+
+# Wyzanczenie unikatowych par współrzednych ze zbioru Nasa i zapisanie w Pandas
+%%time
+distinct_wsp = spark.sql("""
+                          SELECT DISTINCT lon, lat from nasa
+                          """).toPandas()
+
+
+distinct_wsp.shape
+
+
+# Funkcja generująca wykres w czasie wartości zmiennej GVEG - green vegetation dla zadanych współrzednych
+def wykres_szeregu(lon, lat):
+  # ograniczenie zbioru do konkretnej pary współrzędnych
+  szereg = szeregNasa.filter((szeregNasa['lon'] == lon) & (szeregNasa['lat'] == lat)).drop('lon', 'lat')
+  # Przejście na pandas
+  szeregPandas = szereg.toPandas()
+  # Ustawienie 'date' jako indeksu
+  szeregPandas.set_index('Date', inplace=True)
+  # Obliczanie 12-miesięcznej średniej kroczącej
+  szeregPandas['12m_MA'] = szeregPandas['GVEG'].rolling(window=12).mean()
+  # Tworzenie wykresu
+  plt.figure(figsize=(11,6))
+  gs = gridspec.GridSpec(2, 1, height_ratios=[2, 1])
+  xmin, xmax = szeregPandas.index.min() - pd.DateOffset(years=1) , szeregPandas.index.max() + pd.DateOffset(years=1)
+  ax1 = plt.subplot(gs[0])
+  ax1.plot(szeregPandas.index, szeregPandas['GVEG']) #'LAI' - do rozważenia - zmienna opisująca wskaźnik pokrycia liścio
+  ax1.plot(szeregPandas.index, szeregPandas['12m_MA'], label='12-miesięczna średnia krocząca', color='red')
+  ax1.set_title(f'Wykres Szeregu Czasowego (lon {lon}, lat {lat})')
+  ax1.legend(loc='upper left')
+  ax1.set_ylabel('Wartość GVEG')
+  ax1.set_xlim(xmin, xmax)
+  ax1.grid(True)
+
+  ax2 = plt.subplot(gs[1])
+  ax2.plot(szeregPandas.index, szeregPandas['12m_MA'], label='12-miesięczna średnia krocząca', color='red')
+  ax2.set_xlabel('Data')
+  ax2.grid(True)
+  ax2.set_xlim(xmin, xmax)
+
+  plt.tight_layout()
+  plt.show()
```

Comment:
To brzmi jak nowa osobna  funkcja. Nawet tez podobna do poprzedniej. Napiszmy ją jawnie i podobnie ją opiszmy.

Reply:
Chyba nie do końca rozumiem co ma Pan na myśli.